File: api/src/functions/feedback.ts
import { app, HttpRequest, HttpResponseInit, InvocationContext } from '@azure/functions'
import { getAllByTroop, getById, create, update, remove, queryItems } from '../cosmosdb.js'
import { getTroopContext, unauthorized, forbidden } from '../middleware/auth.js'
import { checkPermission } from '../middleware/roles.js'
import { createFeedbackSchema, updateFeedbackSchema, validationError } from '../schemas.js'
import { canViewModeratedContent, moderateTextFields } from '../middleware/moderation.js'

const CONTAINER = 'feedback'

async function feedbackHandler(req: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
  const id = req.params.id
  const method = req.method
  context.log(`${method} /api/feedback${id ? '/' + id : ''}`)

  const auth = await getTroopContext(req, context)
  if (!auth) return unauthorized()

  try {
    if (method === 'GET' && !id) {
      const feedback = await getAllByTroop(CONTAINER, auth.troopId)
      return { jsonBody: feedback.filter((entry) => canViewModeratedContent(auth.role, entry.moderation)) }
    }

    if (method === 'POST') {
      if (!checkPermission(auth.role, 'submitFeedback')) return forbidden()
      const parsed = createFeedbackSchema.safeParse(await req.json())
      if (!parsed.success) return validationError(parsed.error)
      const now = Date.now()
      const audit = { userId: auth.userId, displayName: auth.displayName }
      const moderation = await moderateTextFields([
        { field: 'comments', text: parsed.data.comments },
        { field: 'whatWorked', text: parsed.data.whatWorked },
        { field: 'whatToChange', text: parsed.data.whatToChange },
      ], context)
      const feedback = await create(CONTAINER, {
        id: crypto.randomUUID(),
        troopId: auth.troopId,
        ...parsed.data,
        moderation,
        createdAt: now,
        updatedAt: now,
        createdBy: audit,
        updatedBy: audit,
      })
      return { status: 201, jsonBody: feedback }
    }

    if (method === 'PUT' && id) {
      if (!checkPermission(auth.role, 'submitFeedback')) return forbidden()
      const parsed = updateFeedbackSchema.safeParse(await req.json())
      if (!parsed.success) return validationError(parsed.error)
      const existing = await getById(CONTAINER, id, auth.troopId)
      if (!existing) return { status: 404, jsonBody: { error: 'Feedback not found' } }
      const moderation = await moderateTextFields([
        { field: 'comments', text: parsed.data.comments },
        { field: 'whatWorked', text: parsed.data.whatWorked },
        { field: 'whatToChange', text: parsed.data.whatToChange },
      ], context)
      const feedback = await update(CONTAINER, id, {
        ...existing,
        ...parsed.data,
        id,
        troopId: auth.troopId,
        moderation,
        updatedAt: Date.now(),
        updatedBy: { userId: auth.userId, displayName: auth.displayName },
      }, auth.troopId)
      return { jsonBody: feedback }
    }

    if (method === 'DELETE' && id) {
      if (!checkPermission(auth.role, 'manageEvents')) return forbidden()
      await remove(CONTAINER, id, auth.troopId)
      return { status: 204 }
    }

    return { status: 405, jsonBody: { error: 'Method not allowed' } }
  } catch (err) {
    context.error(`${method} /api/feedback${id ? '/' + id : ''} failed:`, err)
    return { status: 500, jsonBody: { error: 'Internal server error' } }
  }
}

async function feedbackByEventHandler(req: HttpRequest, context: InvocationContext): Promise<HttpResponseInit> {
  const eventId = req.params.eventId
  context.log(`GET /api/feedback/event/${eventId}`)

  const auth = await getTroopContext(req, context)
  if (!auth) return unauthorized()

  try {
    const feedback = await queryItems(
      CONTAINER,
      'SELECT * FROM c WHERE c.eventId = @eventId AND c.troopId = @troopId',
      [
        { name: '@eventId', value: eventId },
        { name: '@troopId', value: auth.troopId },
      ]
    )
    return { jsonBody: feedback.filter((entry: any) => canViewModeratedContent(auth.role, entry.moderation)) }
  } catch (err) {
    context.error(`GET /api/feedback/event/${eventId} failed:`, err)
    return { status: 500, jsonBody: { error: 'Internal server error' } }
  }
}

app.http('feedback', {
  methods: ['GET', 'POST', 'PUT', 'DELETE'],
  authLevel: 'anonymous',
  route: 'feedback/{id?}',
  handler: feedbackHandler,
})

app.http('feedbackByEvent', {
  methods: ['GET'],
  authLevel: 'anonymous',
  route: 'feedback/event/{eventId}',
  handler: feedbackByEventHandler,
})

File: api/app/routers/feedback.py
from __future__ import annotations

import logging
import time
import uuid

from fastapi import APIRouter, HTTPException
from fastapi import Query

from app.cosmosdb import query_items_paginated, get_all_by_troop, get_by_id, create_item, update_item, delete_item, query_items
from app.middleware.auth import RequireTroopContext, forbidden
from app.middleware.roles import check_permission
from app.middleware.moderation import moderate_text_fields, can_view_moderated_content, ModerationField
from app.schemas import CreateFeedback, UpdateFeedback

logger = logging.getLogger(__name__)
router = APIRouter()

CONTAINER = "feedback"
EVENTS_CONTAINER = "events"


@router.get("/feedback")
async def list_feedback(
    auth: RequireTroopContext,
    limit: int = Query(default=50, ge=1, le=100),
    continuationToken: str | None = None,
):
    query = "SELECT * FROM c WHERE c.troopId = @troopId"
    if auth.role != "troopAdmin":
        query += ' AND (NOT IS_DEFINED(c.moderation.status) OR c.moderation.status = "approved")'

    feedback, next_token = await query_items_paginated(
        CONTAINER,
        query,
        [{"name": "@troopId", "value": auth.troopId}],
        limit=limit,
        continuation_token=continuationToken,
    )
    return {"items": feedback, "continuationToken": next_token}


@router.post("/feedback", status_code=201)
async def create_feedback(body: CreateFeedback, auth: RequireTroopContext):
    if not check_permission(auth.role, "submitFeedback"):
        forbidden()
    now = int(time.time() * 1000)
    audit = {"userId": auth.userId, "displayName": auth.displayName}
    moderation = await moderate_text_fields([
        ModerationField(field="comments", text=body.comments),
        ModerationField(field="whatWorked", text=body.whatWorked),
        ModerationField(field="whatToChange", text=body.whatToChange),
    ])
    feedback = await create_item(CONTAINER, {
        "id": str(uuid.uuid4()),
        "troopId": auth.troopId,
        **body.model_dump(),
        "moderation": moderation.__dict__,
        "createdAt": now,
        "updatedAt": now,
        "createdBy": audit,
        "updatedBy": audit,
    })
    return feedback


@router.put("/feedback/{feedback_id}")
async def update_feedback(feedback_id: str, body: UpdateFeedback, auth: RequireTroopContext):
    if not check_permission(auth.role, "submitFeedback"):
        forbidden()
    existing = await get_by_id(CONTAINER, feedback_id, auth.troopId)
    if not existing:
        raise HTTPException(status_code=404, detail="Feedback not found")
    created_by = existing.get("createdBy")
    created_by_user = created_by["userId"] if isinstance(created_by, dict) else ""
    if created_by_user != auth.userId and not check_permission(auth.role, "manageEvents"):
        forbidden("You can only edit your own feedback")
    moderation = await moderate_text_fields([
        ModerationField(field="comments", text=body.comments),
        ModerationField(field="whatWorked", text=body.whatWorked),
        ModerationField(field="whatToChange", text=body.whatToChange),
    ])
    feedback = await update_item(CONTAINER, feedback_id, {
        **existing,
        **body.model_dump(),
        "id": feedback_id,
        "troopId": auth.troopId,
        "moderation": moderation.__dict__,
        "updatedAt": int(time.time() * 1000),
        "updatedBy": {"userId": auth.userId, "displayName": auth.displayName},
    }, auth.troopId)
    return feedback


@router.delete("/feedback/{feedback_id}", status_code=204)
async def delete_feedback(feedback_id: str, auth: RequireTroopContext):
    if not check_permission(auth.role, "manageEvents"):
        forbidden()
    await delete_item(CONTAINER, feedback_id, auth.troopId)


@router.get("/feedback/event/{event_id}")
async def feedback_by_event(event_id: str, auth: RequireTroopContext):
    feedback = await query_items(
        CONTAINER,
        "SELECT * FROM c WHERE c.eventId = @eventId AND c.troopId = @troopId",
        [
            {"name": "@eventId", "value": event_id},
            {"name": "@troopId", "value": auth.troopId},
        ],
    )
    return [f for f in feedback if can_view_moderated_content(auth.role, f.get("moderation"))]


@router.get("/feedback/recipe/{recipe_id}")
async def feedback_by_recipe(recipe_id: str, auth: RequireTroopContext):
    feedback = await query_items(
        CONTAINER,
        "SELECT * FROM c WHERE c.recipeId = @recipeId AND c.troopId = @troopId",
        [
            {"name": "@recipeId", "value": recipe_id},
            {"name": "@troopId", "value": auth.troopId},
        ],
    )
    visible = [f for f in feedback if can_view_moderated_content(auth.role, f.get("moderation"))]

    events = await get_all_by_troop(EVENTS_CONTAINER, auth.troopId)
    events_by_id = {e["id"]: e for e in events}

    return [
        {
            **entry,
            "eventName": events_by_id.get(entry.get("eventId", ""), {}).get("name"),
            "eventDate": events_by_id.get(entry.get("eventId", ""), {}).get("startDate"),
        }
        for entry in visible
    ]
